src/uu/ls/src/url_escaper.rs

```diff
@@ -0,0 +1,145 @@
+use std::{
+    borrow::Cow,
+    convert::{TryFrom, TryInto},
+    ffi::OsStr,
+};
+
+use ascii::{AsciiChar, AsciiStr, AsciiString};
+/// this is logically enum ByteOrResult {Char(AsciiChar), Escape([AsciiChar; 2])}
+/// but we don't use that definition because rustc uses 3 bytes instead of 2 for it (even if we use a pair instead of the array)
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct AsciiOrEscape(Option<AsciiChar>, AsciiChar);
+
+impl AsciiOrEscape {
+    pub fn ascii(ch: AsciiChar) -> Self {
+        Self(None, ch)
+    }
+
+    pub fn escape(c0: AsciiChar, c1: AsciiChar) -> Self {
+        Self(Some(c0), c1)
+    }
+
+    #[inline]
+    pub fn classify(self) -> Result<AsciiChar, [AsciiChar; 2]> {
+        if let Some(c0) = self.0 {
+            Err([c0, self.1])
+        } else {
+            Ok(self.1)
+        }
+    }
```

Comment:
The use of `Result` irks me a bit, because it's not a success/fail scenario, so technically correct, but semantically it doesn't feel right. Maybe use a custom enum?